content/posts/2025-11-06-reducting-microservices/index.md

```diff
@@ -0,0 +1,156 @@
+
+However, these metrics had opposite relationships to decommissioning probability. For Total Cost, higher values directly indicated candidates 
+for removal - expensive services with low returns were prime targets. For the Usage and Maintenance metrics, the logic was inverted: 
+higher values indicated a healthy, actively-used service that should not be decommissioned. Therefore, we applied `1 - normalized_value` 
+to these three metrics, ensuring that low activity translated to high decommissioning scores. 
+This inversion was critical - a service with minimal traffic and few code changes should score high for removal, while a high-traffic, actively 
+maintained service should score low.
+
+We then applied the following score for each metric:
+- Total Cost: 30%
+- \# PRs merged: 20%
+- \# of web requests received: 30%
+- \# of messages processed: 20%
+
+We combined all costs into a single metric because our main goal is service usage rather than cost reduction.
+
+Finally, we applied the following decommissioning score formula for each service:
+
+```
+Decommissioning Score = (0.30 × Total Cost) + (0.20 × # PRs merged) + (0.30 × # of web requests received) + (0.20 * # of messages processed).
+```
+
+We defined a score greater than 80 as indicating a high likelihood of decommissioning the service. 
+A score greater than 50 suggests that further investigation is warranted, while scores below that threshold are not considered significant.
+
+## Execution: From Analysis to Action
+
+The scoring system identified 8% of candidate services as highly likely, with 44% warranting further investigation.
+
+Even after applying the initial score as a filter, a critical analysis was still lacking: **product features in those services**. 
+Is the feature that the service is supposed to deliver still in use? 
+Is it still relevant for our customers? Do we have any plans to leverage it in the future?
+
+We engaged in various research activities to collect insights from Product Managers and Stakeholders. 
+Additionally, a thorough technical assessment of the service was conducted and properly documented. 
+This process eliminated some more services, resulting in 16 out of 45 services identified for decommissioning.
+
+We implemented the following strategy to decommission the remaining services:
+- For services with valuable functionality, we migrated the logic to the appropriate services or libraries.
+- For deprecated services:
+    - First, we added a feature flag on the clients to allow easy activation or deactivation.
+    - After a couple of weeks with no usage and no complaints, we removed the client code.
+    - We created a snapshot of the service’s database.
+    - We shut down all cloud resources associated with the service.
+    - Finally, we wrote thorough documentation explaining the reasons for decommissioning the service, focusing on the assumptions made during the process.
+
+### Results
```

Comment:
Since this will be a public article, I would recommend making this section more generic to avoid exposing anything we don’t intend to share. For example:

> We have decommissioned x% of the microservices supporting the Loadsmart Brokerage Business.
> In terms of cost savings, we estimated a reduction of x% in cloud costs and x hours less of engineering work for maintenance. Considering the investment made, we expect a payback in less than x months.

It might be better to avoid using specific numbers or internal metrics, and instead keep the statements high-level. This helps maintain confidentiality while still communicating the overall impact.